ipaserver/install/ipa_migrate.py

```diff
@@ -0,0 +1,2028 @@
+Database Migration
+--------------------------------
+- Uses a map object (DB_OBJECTS) to identify entries and keep track of what is
+updated
+- First we skip entries that in the excluded list
+- The entry "type" is determined by DB_OBJECTS mapping.  If the type is
+  unknown then the entry is skipped.
+- Skip remote server CA certificate
+    - There migth be a case to keep these, but not as the main CA. TODO discuss
+- Skip the remote "computer"
+- Then the remote entry is cleaned -->  clean_entry()
+    - Remove attributes from the ignore/strip list
+    - Replace suffix/realm/hostname/domainname in all attribute values and DN
+    - Remove objectclasses from groups (IGNORE_OC list)  Might not be needed
+    - userCertificate is removed -> if issued by IPA
+    - Remove unused objectclasses
+- The entry is then checked if it exists on local server.  If it does not exist
+  it is added, otherwise we compare the remote and local entries and update the
+  local entry --> update_local_entry()
+    - Normalize attribute names to match the case of the local server's attrs
+    - Loop over remote entry attributes
+        - Skipping attrs from the "ignore list"
+        - Check the migration mode (prod-mode & stage-mode)
+            - If prod-mode, we migrate SIDs and DNA ranges
+            - If stage-mode, SIDs and DNA are skipped, and dna attributes
+              (uidNumber, gidNumber) are reset to the magic value
+        - Check if attribute is defined in the mappings "special_attrs" list
+            - If it is a special attribute then handle attribute comparison
+              according to the special definition (e.g. list) and update the
+              local entry. Then move on to next attribute...
+        - If the attribute is not "special" then we simply compare attribute to
+          attribute.
+            - If requested, DNA values are reset (magic regen) at this stage
+            - If attribute being updated is "single-valued" then "replace" the
+              value.  If its "multi-valued" then "append" the different value.
+
+Other
+--------------------------------
+There is a lot of normalization going on because we use dictionaries for
+attribute names, but attribute names are CIS, we have to normalize them during
+comparison, but we need to continue to use the original case when we go to
+update the local entry.  So in some cases we normalize to all lowercase, but
+when updatng the local entry we normalize the case of the remote attributes to
+match the local entry's attribute case
+
+
+What's next
+ - ask trivino to skip teardown on CI so I can get data
+
+ Some users/customers add their own entries to the db. Need more info on this
+ as those entries will not be migrated by default
+
+ after password change (but doesn't look like we can reset admin pass:
+     kinit_as_user -> for admin  in IPA API somewhere
+
+ write test from "integration" tests (new class)
+"""
+
+
+logger = logging.getLogger(__name__)
+
+VERSION = "1.0"
```

Comment:
IPA typically doesn't independently version tools. Is there some benefit to split the version from the tag?

Reply:
> IPA typically doesn't independently version tools. Is there some benefit to split the version from the tag?

There is no real reason for this, I'll remove it